### PR TITLE
apk: Support replacing existing debuggable attribute

### PR DESCRIPTION
**Pull request by tchebb**
If a package explicitly specifies debuggable="false", adding a new entry setting it to "true" has no effect. To properly patch such APKs, we need to replace the entry.

This contribution is on behalf of my company.

### CODE REVIEW

**Comment by s1341** [comment]
Has this been tested?

**Comment by tchebb** [comment]
Yes. I've tested it on both the Facebook app (`com.facebook.katana.456014858.apk` as of today, downloadable from [here](https://www.facebook.com/android_upgrade)) and the Magisk app ([`Magisk-v28.0.apk`](https://github.com/topjohnwu/Magisk/releases/download/v28.0/Magisk-v28.0.apk)). frida-apk works on Magisk both before and after this PR, while it only works on Facebook after this PR.

Here are how the apps' manifests look on the unmodified APKs:
```
$ aapt d xmltree com.facebook.katana.456014858.apk AndroidManifest.xml | grep android:debuggable
      A: android:debuggable(0x0101000f)=(type 0x12)0x0
$ aapt d xmltree Magisk-v28.0.apk AndroidManifest.xml | grep android:debuggable
<no output>
```

Here's how they look after frida-apk, without this PR (note the duplicate entry on the Facebook APK):
```
$ aapt d xmltree com.facebook.katana.456014858.d-current.apk AndroidManifest.xml | grep android:debuggable
      A: android:debuggable(0x0101000f)=(type 0x12)0x0
      A: android:debuggable(0x0101000f)=(type 0x12)0xffffffff
$ aapt d xmltree Magisk-v28.0.d-current.apk AndroidManifest.xml | grep android:debuggable
      A: android:debuggable(0x0101000f)=(type 0x12)0xffffffff
```

And here's how they look after frida-apk, with this PR:
```
$ aapt d xmltree com.facebook.katana.456014858.d-new.apk AndroidManifest.xml | grep android:debuggable
      A: android:debuggable(0x0101000f)=(type 0x12)0xffffffff
$ aapt d xmltree Magisk-v28.0.d-new.apk AndroidManifest.xml | grep android:debuggable
      A: android:debuggable(0x0101000f)=(type 0x12)0xffffffff
```

On my Pixel 5 running Android 14, `xmltree com.facebook.katana.456014858.d-new.apk` shows up as debuggable when installed, while `com.facebook.katana.456014858.d-current.apk` doesn't.

File: frida_tools/apk.py [hunk 1]
```diff
@@ -191,18 +191,26 @@ def insert_debuggable(self, name: int, resource_map: ResourceMap) -> None:
...
         attr_offset = None
+        replace = False
         for insert_pos in range(self.attribute_count + 1):
-            attr_offset = 0x24 + 20 * insert_pos
+            attr_offset = 0x24 + struct.calcsize(self.ATTRIBUTE_FORMAT) * insert_pos
             idx = int.from_bytes(chunk_data[attr_offset + 4 : attr_offset + 8], "little")
-            if resource_map.get_resource(idx) > ResourceMap.DEBUGGING_RESOURCE:
+            res = resource_map.get_resource(idx)
+            if res >= ResourceMap.DEBUGGING_RESOURCE:
+                # If there's already a debugging resource, replace it.
```
Comment (oleavr) [inline frida_tools/apk.py:200]:
```suggestion
```

Code already makes this clear.

File: frida_tools/apk.py [hunk 2]
```diff
@@ -191,18 +191,26 @@ def insert_debuggable(self, name: int, resource_map: ResourceMap) -> None:
 
         # Some parts of Android expect this to be sorted by resource ID.
         attr_offset = None
+        replace = False
         for insert_pos in range(self.attribute_count + 1):
-            attr_offset = 0x24 + 20 * insert_pos
+            attr_offset = 0x24 + struct.calcsize(self.ATTRIBUTE_FORMAT) * insert_pos
```
Comment (oleavr) [inline frida_tools/apk.py:196]:
```suggestion
            attr_offset = 0x24 + insert_pos * struct.calcsize(self.ATTRIBUTE_FORMAT)
```

Nit: Existing code has the quantity first, size second.

**Review by oleavr — CHANGES_REQUESTED** [review]
Thanks for doing this! :raised_hands:

**Pull request merged**